craft_providers/actions/snap_installer.py

```diff
@@ -0,0 +1,191 @@
+        raise SnapInstallationError(
+            brief="Unable to connect to snapd service."
+        ) from error
+
+    try:
+        resp.raise_for_status()
+    except requests.exceptions.HTTPError as error:
+        raise SnapInstallationError(
+            brief=f"Unable to download snap {snap_name!r} from snapd."
+        ) from error
+
+    with output.open("wb") as stream:
+        for chunk in resp.iter_content(chunk_size):
+            stream.write(chunk)
+
+
+def _pack_host_snap(*, snap_name: str, output: pathlib.Path) -> None:
+    """Pack the current host snap."""
+    cmd = [
+        "snap",
+        "pack",
+        f"/snap/{snap_name}/current/",
+        f"--filename={output.as_posix()}",
+    ]
+
+    logger.debug("Executing command on host: %s", shlex.join(cmd))
+    subprocess.run(
+        cmd,
+        capture_output=True,
+        check=True,
+    )
+
+
+@contextlib.contextmanager
+def _get_host_snap(snap_name: str) -> Iterator[pathlib.Path]:
+    """Get snap installed on host.
+
+    Snapd provides an API to fetch a snap. First use that to fetch a snap.
+    If the snap is installed using `snap try`, it may fail to download. In
+    that case, attempt to construct the snap by packing it ourselves.
+
+    :yields: Path to snap which will be cleaned up afterwards.
+    """
+    with temp_paths.home_temporary_directory() as tmp_dir:
+        snap_path = tmp_dir / f"{snap_name}.snap"
+
+        try:
+            _download_host_snap(snap_name=snap_name, output=snap_path)
+        except SnapInstallationError:
+            logger.warning(
+                "Failed to fetch snap from snapd, falling back to `snap pack` to recreate."
+            )
+            _pack_host_snap(snap_name=snap_name, output=snap_path)
+
+        yield snap_path
+
+
+def inject_from_host(*, executor: Executor, snap_name: str, classic: bool) -> None:
+    """Inject snap from host snap.
+
+    :raises SnapInstallationError: on unexpected error.
+    """
+    target_snap_path = pathlib.Path(f"/tmp/{snap_name}.snap")
+
+    try:
+        # Clean outdated snap, if exists.
+        executor.execute_run(
+            ["rm", "-f", target_snap_path.as_posix()],
+            check=True,
+            capture_output=True,
+        )
+
+        with _get_host_snap(snap_name) as host_snap_path:
+            executor.push_file(
+                source=host_snap_path,
+                destination=target_snap_path,
+            )
+
+        executor.execute_run(
+            snap_cmd.formulate_install_command(
+                classic=classic, dangerous=True, snap_path=target_snap_path
+            ),
+            check=True,
+            capture_output=True,
+        )
+    except subprocess.CalledProcessError as error:
+        raise SnapInstallationError(
+            brief=f"Failed to inject snap {snap_name!r}.",
+            details=details_from_called_process_error(error),
+        ) from error
+    except ProviderError as error:
```

Comment:
Remind me what can raise a ProviderError here? Would it not be relevant to add "details"?

Reply:
push_file() will raise a ProviderError for whatever error may be raised during the file copy.  It shouldn't happen normally, but it would contain the reason (as perhaps explained by multipass/lxd command output).   I chain it so the data is available if necessary.

It would be clearer if I didn't combine the two actions, so I'll do that and add a detail there.

Reply:
Done & added a test.